shared_model/interfaces/common_objects/common_objects_factory.hpp

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_COMMON_OBJECTS_FACTORY_HPP
+#define IROHA_COMMON_OBJECTS_FACTORY_HPP
+
+#include <memory>
+
+#include "common/result.hpp"
+#include "interfaces/common_objects/account.hpp"
+#include "interfaces/common_objects/peer.hpp"
+#include "interfaces/common_objects/types.hpp"
+#include "interfaces/common_objects/domain.hpp"
+
+namespace shared_model {
+  namespace interface {
+    /**
+     * CommonObjectsFactory provides methods to construct simple objects
+     * such as peer, account etc.
+     */
+    class CommonObjectsFactory {
+     public:
+      template <typename T>
+      using FactoryResult = iroha::expected::Result<T, std::string>;
+
+      /**
+       * Create peer instance
+       */
+      virtual FactoryResult<std::unique_ptr<Peer>> createPeer(
+          const types::AddressType &address,
+          const types::PubkeyType &public_key) = 0;
+
+      /**
+       * Create account instance
+       */
+      virtual FactoryResult<std::unique_ptr<Account>> createAccount(
+          const types::AccountIdType &account_id,
+          const types::DomainIdType &domain_id,
+          types::QuorumType quorum,
+          const types::JsonType &jsonData) = 0;
+
+      /**
+       * Create account asset instance
+       */
+      virtual FactoryResult<std::unique_ptr<AccountAsset>> createAccountAsset(
+          const types::AccountIdType &account_id,
+          const types::AssetIdType &asset_id,
+          const Amount &balance) = 0;
+
+      /**
+       * Create amount instance from string
+       *
+       * @param value integer will be divided by 10 * precision,
+       * so value 123 with precision 2 will become Amount of 1.23
+       */
+      virtual FactoryResult<std::unique_ptr<Amount>> createAmount(
+          boost::multiprecision::uint256_t value,
+          types::PrecisionType precision) = 0;
+
+      /**
+       * Create amount instance from string
+       *
+       * @param amount must represent valid number.
+       * For example: "1.23", "10" etc.
+       */
+      virtual FactoryResult<std::unique_ptr<Amount>> createAmount(
+          std::string amount) = 0;
```

Comment:
Why not const reference?

Reply:
because we modify this parameter in the code below. This logic was not written by me, and will be thrown out in the future, so it is simpler to just pass by value for now.